src/main/java/teammates/ui/controller/InstructorFeedbackResultsPageData.java

```diff
@@ -226,266 +226,189 @@ public void initForSectionPanelViews(InstructorAttributes instructor,
-                    } else {
-                        buildMissingParticipantPanelsForSection(sectionPanel, prevSection, participantsWithResponses);
-                        participantsWithResponses.clear();
-                    }
-                    
-                    // add to sectionPanels,
-                    sectionPanels.put(prevSection, sectionPanel);
-                    sectionsWithResponses.add(prevSection);
-                    
-                    // setup for next section
-                    teamsWithResponses.clear();
-                    teamsWithResponses.add(currentTeam);
-                    
-                    sectionPanel = new InstructorFeedbackResultsSectionPanel();
-                }
-            }
-            
-            if (isDifferentTeam) {
                 sectionPanel.getIsTeamWithResponses().put(currentTeam, true);
             }
-            
+
             // Build participant panel for the current primary participant
-            InstructorFeedbackResultsParticipantPanel primaryParticipantPanel =
-                    buildGroupByParticipantPanel(primaryParticipantIdentifier,
-                                                 primaryToSecondaryParticipantToResponsesMap,
-                                                 additionalInfoId, primaryParticipantIndex);
-            
-            if (isGroupedByTeam()) {
-                sectionPanel.addParticipantPanel(currentTeam, primaryParticipantPanel);
-            } else {
-                sectionPanel.addParticipantPanel(primaryParticipantPanel);
-            }
-            
-            participantsWithResponses.add(primaryParticipantIdentifier);
-            
+            InstructorFeedbackResultsParticipantPanel recipientPanel =
+                    buildGroupByParticipantPanel(primaryParticipantIdentifier, primaryToSecondaryParticipantToResponsesMap,
+                                                 sectionPrefix + additionalInfoId, primaryParticipantIndex);
+
+            sectionPanel.addParticipantPanel(currentTeam, recipientPanel);
+
+            teamMembersWithResponses.add(primaryParticipantIdentifier);
+
             prevTeam = currentTeam;
-            prevSection = currentSection;
         }
-        
+
         // for the last section with responses
-        if (isGroupedByTeam()) {
-            buildMissingParticipantPanelsForTeam(sectionPanel, prevTeam, participantsWithResponses);
-        }
-        
+        buildMissingParticipantPanelsForTeam(sectionPanel, prevTeam, teamMembersWithResponses);
+
         teamsWithResponses.add(prevTeam);
-        if (isGroupedByTeam()) {
-            buildMissingTeamAndParticipantPanelsForSection(
-                                            sectionPanel, prevSection, teamsWithResponses);
-        } else {
-            buildMissingParticipantPanelsForSection(sectionPanel, prevSection, participantsWithResponses);
```

Comment:
For future developers' reference, here are the settings to reproduce the incorrect missing participants, as discussed:

![class_roster](https://cloud.githubusercontent.com/assets/2216119/16043607/074555bc-3273-11e6-8aac-1f494ba02819.PNG)
(The demo course students, but renamed "Alice" to "Jenny", and added an additional team whose student "Jean" < "Jenny".)

Use a non-question, but without grouping by team:
![results_filter](https://cloud.githubusercontent.com/assets/2216119/16043685/82a816ae-3273-11e6-86b1-81e0bed9ff3f.PNG)

Delete all responses from emma: this causes the "missing participants" logic to generate missing responses for emma' entire team
![incorrect_missing_participant](https://cloud.githubusercontent.com/assets/2216119/16043666/65d4fa6a-3273-11e6-91cf-8dfc4458f003.PNG)

Expected result:
![master](https://cloud.githubusercontent.com/assets/2216119/16043836/423fe320-3274-11e6-826f-c90f36b250d4.PNG)

@belyabl9 sorry I have not yet looked through the existing test cases as promised. I will try to do it soon.